dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -535,82 +533,56 @@ ColumnPtr DMFileReader::readColumn(const ColumnDefine & cd, size_t start_pack_id
     }
 
+    auto data_type = dmfile->getColumnStat(cd.id).type;
+
     // Not cached
     if (!enable_column_cache || !isCacheableColumn(cd))
     {
-        auto data_type = dmfile->getColumnStat(cd.id).type;
-        ColumnPtr column;
-        readFromDiskOrSharingCache(cd, column, start_pack_id, pack_count, read_rows);
+        auto column = readFromDiskOrSharingCache(cd, start_pack_id, pack_count, read_rows);
         return convertColumnByColumnDefineIfNeed(data_type, std::move(column), cd);
     }
 
     // enable_column_cache && isCacheableColumn(cd)
-    auto read_strategy = column_cache->getReadStrategy(start_pack_id, pack_count, cd.id);
-    const auto & pack_stats = dmfile->getPackStats();
-    auto data_type = dmfile->getColumnStat(cd.id).type;
-    auto column = data_type->createColumn();
-    column->reserve(read_rows);
-    for (auto & [range, strategy] : read_strategy)
-    {
-        if (strategy == ColumnCache::Strategy::Memory)
-        {
-            for (size_t cursor = range.first; cursor < range.second; cursor++)
-            {
-                auto cache_element = column_cache->getColumn(cursor, cd.id);
-                column->insertRangeFrom(
-                    *(cache_element.first),
-                    cache_element.second.first,
-                    cache_element.second.second);
-            }
-        }
-        else if (strategy == ColumnCache::Strategy::Disk)
-        {
-            size_t rows_count = 0;
-            for (size_t cursor = range.first; cursor < range.second; cursor++)
-            {
-                rows_count += pack_stats[cursor].rows;
-            }
-            ColumnPtr col;
-            readFromDiskOrSharingCache(cd, col, range.first, range.second - range.first, rows_count);
-            column->insertRangeFrom(*col, 0, col->size());
-        }
-        else
-        {
-            throw Exception("Unknown strategy", ErrorCodes::LOGICAL_ERROR);
-        }
-    }
-    ColumnPtr result_column = std::move(column);
-    size_t rows_offset = 0;
-    for (size_t cursor = start_pack_id; cursor < start_pack_id + pack_count; cursor++)
-    {
-        column_cache->tryPutColumn(cursor, cd.id, result_column, rows_offset, pack_stats[cursor].rows);
-        rows_offset += pack_stats[cursor].rows;
-    }
+
+    // try to get column from cache
+    MutableColumnPtr result_column = data_type->createColumn();
+    bool success = getColumnFromCache(
+        column_cache,
+        cd,
+        start_pack_id,
+        pack_count,
+        read_rows,
+        result_column,
+        [&](const ColumnDefine & cd, size_t start_pack_id, size_t pack_count, size_t read_rows) {
+            return readFromDiskOrSharingCache(cd, start_pack_id, pack_count, read_rows);
+        });
+    RUNTIME_CHECK(success);
+    // add column to cache
+    ColumnPtr column = std::move(result_column);
+    addColumnToCache(column_cache, cd.id, start_pack_id, pack_count, column);
```

Comment:
I didn't get it why `getColumnFromCache` should always success here.

And in the last line of `getColumnFromCache`, it will delete the column from `column_cache`. But line 562 will add the column into `column_cache` again. Is it reasonable?

Reply:
ok